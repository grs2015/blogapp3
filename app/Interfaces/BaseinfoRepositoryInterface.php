<?php

namespace App\Interfaces;

use Illuminate\Support\Collection;
use Illuminate\Database\Eloquent\Model;

interface BaseinfoRepositoryInterface
{
    public function getAllEntries():?Collection;
    public function getEntryById(int $baseId):?Model;
    public function deleteEntry(int $baseId):int;
    public function createEntry(array $baseAttributes):?Model;
    public function updateEntry(int $baseId, array $newAttributes):?int;
}
