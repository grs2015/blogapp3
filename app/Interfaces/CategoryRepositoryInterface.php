<?php

namespace App\Interfaces;

use Illuminate\Support\Collection;
use Illuminate\Database\Eloquent\Model;

interface CategoryRepositoryInterface
{
    public function getAllEntries():?Collection;
    public function getEntryById(int $catId):?Model;
    public function deleteEntry(int $catId):int;
    public function createEntry(array $catAttributes):?Model;
    public function updateEntry(int $catId, array $newAttributes):?int;
}
