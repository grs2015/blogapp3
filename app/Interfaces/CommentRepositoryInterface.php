<?php

namespace App\Interfaces;

use Illuminate\Support\Collection;
use Illuminate\Database\Eloquent\Model;

interface CommentRepositoryInterface
{
    public function getAllEntries(int $postId):?Collection;
    public function getEntryById(int $postId, int $commentId):?Model;
    public function deleteEntry(int $postId, int $commentId):void;
    public function createEntry(int $postId, array $commentAttributes):?Model;
    public function updateEntry(int $postId, int $commentId, array $newAttributes):?int;
}
