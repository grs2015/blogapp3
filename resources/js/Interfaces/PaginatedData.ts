export interface Paginated {
    'current_page': number;
    'data': Array<PostData>;
    'first_page_url': string;
    'from': number;
    'last_page': number;
    'last_page_url': string;
    'links': Array<LinkData>;
    'next_page_url': string | null;
    'path': string;
    'per_page': number;
    'prev_page_url': string | null;
    'to': number;
    'total': number;
}

export interface PostData {
    'author_id'?: number | null;
    'content'?: string | null;
    'favorite': Favorite;
    'hero_image'?: File | any | any[] | null,
    'id': number;
    'meta_title'?: string | null;
    'parent_id'?: number | null;
    'published_at'?: string | null;
    'slug'?: string | null;
    'status': Status;
    'summary'?: string | null;
    'time_to_read': number;
    'title': string;
    'user'?: UserData | null;
    'views'?: string | null;
    'galleries'?: Array<GalleryData> | null,
    'postmetas'?: Array<PostmetaData> | null,
    'tag_ids'?: Array<number> | null,
    'cat_ids': Array<number>,
    'categories'?: Array<categoryData> | null,
    'tags'?: Array<tagData> | null,
    'images'?: Array<File> | null,
    '_method'?: "put" | null
}

export interface SortingData {
    column: string;
    descending: string;
}

export interface LinkData {
    'active': boolean;
    'label': string | null;
    'url': string |null;
}

type UserData = {
    'email': string;
    'first_name': string;
}

type PostmetaData = {
    id?: number | null,
    post_id: number,
    key: string,
    content: string
}

export interface GalleryData {
    id: number;
    post_id: number;
    original: string;
    lowres: string;
    thumbs: string;
}

export interface tablePagination {
    'sortBy': string | null,
    'descending': boolean | null,
    'page': number,
    'rowsPerPage': number,
    'rowsNumber': number,
}

export enum Favorite {
    Usual = "usual",
    Favorite = "favorite"
}

export enum Status {
    Published = 'published',
    Draft = 'draft',
    Pending = 'pending',
    Unpublished = 'unpublished',
}

export enum UserStatus {
    Enabled = 'enabled',
    Disabled = 'disabled',
    Pending = 'pending'
}

export enum UserRole {
    Admin = 'admin',
    Author = 'author',
    Member = 'member',
    SuperAdmin = 'super-admin'
}

export interface categoryData {
    id: number,
    title: string,
    meta_title?: string,
    content?: string,
    parent_id?: number,
    slug?: string,
    icon?: string,
    color?: string
}

export interface tagData {
    content?: string,
    id: number,
    meta_title?: string,
    slug: string,
    title: string
}

export interface breadcrumbsData {
    label: string,
    icon: string,
    route?: string
}

export interface userData {
    avatar?: string,
    email: string,
    first_name: string,
    full_name?: string,
    id: number,
    intro?: string,
    last_login?: string,
    last_name?: string,
    middle_name?: string,
    mobile?: string,
    posts_count?: number,
    profile?: string,
    registered_at?: string,
    roles?: string,
    status?: UserStatus,
    email_verified_at?: string,
}

export interface Role {
    created_at: string,
    updated_at: string,
    guard_name: string,
    id: number,
    name: string,
    pivot: Object

}
