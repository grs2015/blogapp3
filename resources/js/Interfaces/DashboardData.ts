import { Status, UserStatus } from "./PaginatedData"

export interface DashboardData extends DashboardPostData {
    active_authors: Array<ActiveUser>,
    count_author_users: number,
    count_member_users: number,
    enabled_users_count: number,
    pending_users_count: number,
    recent_users: Array<RecentUser>
}

export interface DashboardPostData {
    recent_posts: Array<RecentPost>,
    draft_posts_count: number,
    pending_posts_count: number,
    posts_count: number,
    published_posts_count: number,
    most_rated: Array<RatedPost>,
    most_viewed: Array<ViewedPost>,
    most_commented: Array<CommentedPost>
}

export interface UserData {
    email: string,
    full_name: string
}

export interface ActiveUser {
    id: number,
    email: string,
    full_name: string,
    last_login: string,
    posts_count: number,
    roles: string
}
export interface RecentUser {
    id: number,
    created_at: string,
    email: string,
    roles: string,
    full_name: string,
    status: UserStatus,
}
export interface RecentPost {
    id: number,
    published_at: string,
    status: Status,
    title: string,
    user?: UserData
}
export interface RatedPost {
    published_at: string,
    rating: number,
    title: string,
    user?: UserData,
}
export interface ViewedPost {
    published_at: string,
    title: string,
    views: number,
    user?: UserData
}
export interface CommentedPost {
    comments_count: number,
    published_at: string,
    title: string,
    user?: UserData
}
