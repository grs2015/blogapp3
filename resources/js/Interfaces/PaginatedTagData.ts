import { tagData, LinkData } from "./PaginatedData";


export interface PaginatedTag {
    'current_page': number;
    'data': Array<tagData>;
    'first_page_url': string;
    'from': number;
    'last_page': number;
    'last_page_url': string;
    'links': Array<LinkData>;
    'next_page_url': string | null;
    'path': string;
    'per_page': number;
    'prev_page_url': string | null;
    'to': number;
    'total': number;
}
