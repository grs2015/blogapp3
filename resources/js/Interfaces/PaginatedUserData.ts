import { userData, LinkData } from "./PaginatedData";


export interface PaginatedUser {
    'current_page': number;
    'data': Array<userData>;
    'first_page_url': string;
    'from': number;
    'last_page': number;
    'last_page_url': string;
    'links': Array<LinkData>;
    'next_page_url': string | null;
    'path': string;
    'per_page': number;
    'prev_page_url': string | null;
    'to': number;
    'total': number;
}
