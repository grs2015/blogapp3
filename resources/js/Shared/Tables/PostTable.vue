<script setup lang="ts">

import { ref, computed, watch } from 'vue'
import { trans } from 'laravel-vue-i18n';
import { Inertia, PageProps } from '@inertiajs/inertia'
import { PostData, LinkData, tablePagination } from '@/Interfaces/PaginatedData';
import { useQuasar } from 'quasar'
// import { usePage } from '@inertiajs/inertia-vue3';

interface Paginated {
    paginatedData: {
        current_page: number;
        data: Array<PostData>;
        first_page_url: string;
        from: number;
        last_page: number;
        last_page_url: string;
        links: Array<LinkData>;
        next_page_url: string | null;
        path: string;
        per_page: number;
        prev_page_url: string | null;
        to: number;
        total: number;
    },
    sortingData: {
        column: string;
        descending: string;
    }
}

interface pageActions extends PageProps
{
    sorting: {
        column: string;
        descending: string;
    }
}

const props = defineProps<Paginated>()
const $q = useQuasar()

const rows = computed(() => {
    loading.value = false
    return props.paginatedData.data
})
const filter = ref('')
const loading = ref(false)
const loadingResetButton = ref(false)

const pagination = ref<tablePagination>({
    // sortBy: usePage<pageActions>().props.value.sorting.column,
    // descending: usePage<pageActions>().props.value.sorting.descending === 'true',
    sortBy: props.sortingData.column,
    descending: props.sortingData.descending === 'true',
    page: props.paginatedData.current_page,
    rowsPerPage: props.paginatedData.per_page,
    rowsNumber: props.paginatedData.total
})

type tableColumns = {
    name: string;
    label: string;
    field: string | ((row: PostData) => string);
    required?: boolean;
    align?: "left" | "right" | "center";
    sortable?: boolean;
}

const columns:tableColumns[] = [
    {
        name: 'postTitle',
        required: true,
        label: trans('Title'),
        align: 'left',
        sortable: true,
        field: (row : PostData) => row.title,
        style: "width: 30%"
    },
    {
        name: 'postAuthor',
        required: true,
        label: trans('Author'),
        align: 'left',
        field: (row : PostData) => row.user.first_name
    },
    {
        name: 'postStatus',
        required: true,
        label: trans('Status'),
        align: 'right',
        sortable: true,
        field: (row : PostData) => row.status,
    },
    {
        name: 'postFavorite',
        required: true,
        label: trans('Favorite'),
        align: 'right',
        sortable: true,
        field: (row : PostData) => row.favorite
    },
    {
        name: 'postViews',
        required: true,
        label: trans('Views'),
        align: 'right',
        sortable: true,
        field: (row : PostData) => row.views
    },
    {
        name: 'postActions',
        required: true,
        label: trans('Actions'),
        align: 'center',
        field: 'actions'
    }
]

const onRequest = (props: { pagination : tablePagination, filter: string }) => {
    loading.value = true
    const { page, rowsPerPage, sortBy, descending } = props.pagination
    const filter = props.filter
    Inertia.get('/admin/posts', { per_page: rowsPerPage, page: page, [sortBy]: descending ? "asc" : "desc", column: sortBy, descending: descending, search: filter })
    pagination.value.page = page
    pagination.value.rowsPerPage = rowsPerPage
    pagination.value.sortBy = sortBy
    pagination.value.descending = descending
}

const reset = () => {
    loading.value = true
    loadingResetButton.value = true
    Inertia.get('/admin/posts')
}

const editPost = (row: Object) => {
    let postSlug = row['slug']
    Inertia.get(`/admin/posts/${postSlug}/edit`)
}

const deletePost = (row: Object) => {
    let postSlug = row['slug']
    loading.value = true
    Inertia.delete(`/admin/posts/${postSlug}`, {
        onSuccess: () => onDeleteSuccess(),
        onError: () => onDeleteFail()
    })
}

const onDeleteFail = () => {
    $q.notify({
        type: 'negative',
        message: "Something went wrong with post deletion"
    })
}

const onDeleteSuccess = () => {
    $q.notify({
        type: 'positive',
        message: "The post have been deleted successfully"
    })
}

const addPost = () => {
    Inertia.get('/admin/posts/create')
}

</script>

<template>
    <div class="q-mt-md">
        <q-table :title="$t('Posts')" :row-key="row => row.id" :columns="columns" :rows="rows"
            v-model:pagination="pagination" :filter="filter" @request="onRequest" :loading="loading" color="primary"
            flat binary-state-sort :rows-per-page-label="$t('Records per page')"
            :rows-per-page-options="[5, 10, 15, props.paginatedData.total, 0]" bordered>
            <template v-slot:header="props">
                <q-tr :props="props">
                    <q-th v-for="col in props.cols" :key="col.name" :props="props" class="text-primary">
                        {{ col.label }}
                    </q-th>
                </q-tr>
            </template>
            <template v-slot:no-data>
                <div class="full-width row flex-center text-primary q-gutter-sm">
                    <q-icon size="2em" name="sentiment_dissatisfied" color="red" />
                    <span class="text-red">
                        {{ $t('No data available') }}
                    </span>
                </div>
            </template>
            <template v-slot:body-cell-postActions="props">
                <q-td :props="props" auto-width>
                    <div class="row flex-center q-gutter-x-sm no-wrap">
                        <q-btn outline color="accent" icon="edit" :disable="loading" @click="editPost(props.row)" data-test="edit-button">
                            <q-tooltip :delay="1000" anchor="bottom middle" self="center middle">
                                {{ $t('Edit post') }}
                            </q-tooltip>
                        </q-btn>
                        <q-btn outline color="red" icon="delete" :disable="loading" @click="deletePost(props.row)" data-test="delete-button">
                            <q-tooltip :delay="1000" anchor="bottom middle" self="center middle">
                                {{ $t('Delete post') }}
                            </q-tooltip>
                        </q-btn>
                    </div>
                </q-td>
            </template>
            <template v-slot:body-cell-postViews="props">
                <q-td :props="props" auto-width>
                    {{ props.row.views }}
                </q-td>
            </template>
            <template v-slot:body-cell-postFavorite="props">
                <q-td :props="props" auto-width>
                    {{ props.row.favorite }}
                </q-td>
            </template>
            <template v-slot:body-cell-postStatus="props">
                <q-td :props="props" auto-width>
                    {{ props.row.status }}
                </q-td>
            </template>
            <template v-slot:top>
                <div class="q-table__title text-primary">{{ $t('Posts') }}</div>
                <q-space />
                <q-input clear-icon="close" dense debounce="300" label-color="primary" v-model="filter" label-slot data-test="search-input">
                    <template v-slot:append>
                        <q-icon name="search" color="primary" />
                    </template>
                    <template v-slot:label>
                        <div>
                            {{ $t('Search') }}
                        </div>
                    </template>
                </q-input>
                <q-btn class="q-ml-md" color="primary" outline :disable="loading" @click="reset" data-test="refresh-button"
                    :loading="loadingResetButton" icon="restart_alt" />
                <q-separator vertical spaced inset />
                <q-btn color="green" unelevated @click="addPost" data-test="add-button">
                    <q-icon left name="post_add" />
                    <div>{{ $t('ADD POST') }}</div>
                </q-btn>
            </template>
        </q-table>
    </div>
</template>
