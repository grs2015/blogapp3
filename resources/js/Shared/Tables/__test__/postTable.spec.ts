import { mount } from "@vue/test-utils";
import PostTable from "@/Shared/Tables/PostTable.vue"
import { Paginated } from "@/Interfaces/PaginatedData";
import { i18nVue } from "laravel-vue-i18n";
import { Quasar } from "quasar";
import { Favorite, Status } from "@/Interfaces/PaginatedData";
import { Inertia } from '@inertiajs/inertia'

vi.spyOn(Inertia, 'get').mockImplementation(() => console.log('Sending Get request'))
vi.spyOn(Inertia, 'delete').mockImplementation(() => console.log('Sending Delete request'))

let wrapper = null

const paginatedData = {
    current_page: 1,
    data: [
        { id: 1, title: 'first_title', favorite: Favorite.Usual, parent_id: 1, views: '1', slug: 'slug', status: Status.Draft, time_to_read: 2, user: { email: 'email', first_name: 'name' } },
        { id: 2, title: 'first_title', favorite: Favorite.Usual, parent_id: 1, views: '1', slug: 'slug', status: Status.Draft, time_to_read: 2, user: { email: 'email', first_name: 'name' } },
    ],
    first_page_url: "first_page_url",
    from: 1,
    last_page: 4,
    last_page_url: "last_page_url",
    links: [ { active: true, label: 'active', url: 'link' }, { active: false, label: 'inactive', url: 'link' } ],
    next_page_url: "next_page_url",
    path: "path",
    per_page: 5,
    prev_page_url: "prev_page_url",
    to: 4,
    total: 20
}
const sortingData = { column: 'title', descending: 'true' }


beforeEach(async() => {

    wrapper = await mount(PostTable,{ global: {
        plugins: [i18nVue, Quasar] },
        props: { paginatedData, sortingData },
        shallow: false })
    // console.log(wrapper.html())
})
afterEach(async () => await wrapper.unmount());

describe('PostTable component', () => {
    it('checks the data from parent component is defined on current component', () => {
        expect(wrapper.element).toMatchSnapshot();

        expect(wrapper.props().paginatedData).toContain({
            current_page: 1,
            first_page_url: 'first_page_url',
            last_page_url: 'last_page_url',
            prev_page_url: 'prev_page_url',
        })
        expect(wrapper.props().paginatedData.data[0]).toContain({
            title: 'first_title',
            slug: 'slug',
            status: 'draft'
        })
    })

    it('checks the number of row in table (1 - Head, 2 - Body)', () => {
        expect(wrapper.findAll('tr')).toHaveLength(3)
    })

    it('checks the button events correcly hit the Inertia back-end', async () => {
        const editBtn = wrapper.find('[data-test="edit-button"]')
        const deleteBtn = wrapper.find('[data-test="delete-button"]')
        const refreshBtn = wrapper.find('[data-test="refresh-button"]')
        const addBtn = wrapper.find('[data-test="add-button"]')

        await refreshBtn.trigger('click')
        expect(Inertia.get).toHaveBeenCalledTimes(1)
        expect(Inertia.get).toHaveBeenCalledWith("/admin/posts")

        await editBtn.trigger('click')
        expect(Inertia.get).toHaveBeenCalledTimes(2)
        expect(Inertia.get).toHaveBeenCalledWith("/admin/posts/slug/edit")

        await addBtn.trigger('click')
        expect(Inertia.get).toHaveBeenCalledTimes(3)
        expect(Inertia.get).toHaveBeenCalledWith("/admin/posts/create")

        await deleteBtn.trigger('click')
        expect(Inertia.delete).toHaveBeenCalledTimes(1)
        expect(Inertia.delete).toHaveBeenCalledWith("/admin/posts/slug")
    })
})
