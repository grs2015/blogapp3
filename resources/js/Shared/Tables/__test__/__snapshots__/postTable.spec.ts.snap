// Vitest Snapshot v1

exports[`PostTable component > checks the data from parent component is defined on current component 1`] = `
<div
  class="q-mt-md"
>
  <div
    class="q-table__container q-table--horizontal-separator column no-wrap q-table__card q-table--flat q-table--bordered q-table--no-wrap"
  >
    <div
      class="q-table__top relative-position row items-center"
    >
      
      <div
        class="q-table__title text-primary"
      >
        Posts
      </div>
      <div
        class="q-space"
      />
      <label
        class="q-field row no-wrap items-start q-field--standard q-input q-field--labeled q-field--dense"
        for="f_6b8253ab-b2bf-482d-a1c1-2c96fc56e5d4"
      >
        <!---->
        <div
          class="q-field__inner relative-position col self-stretch"
        >
          <div
            class="q-field__control relative-position row no-wrap"
            tabindex="-1"
          >
            <div
              class="q-field__control-container col relative-position row no-wrap q-anchor--skip"
            >
              <input
                class="q-field__native q-placeholder"
                data-test="search-input"
                id="f_6b8253ab-b2bf-482d-a1c1-2c96fc56e5d4"
                tabindex="0"
                type="text"
              />
              <div
                class="q-field__label no-pointer-events absolute ellipsis text-primary"
              >
                <div>
                  Search
                </div>
              </div>
              <!---->
            </div>
            <div
              class="q-field__append q-field__marginal row no-wrap items-center"
            >
              <i
                aria-hidden="true"
                class="q-icon text-primary notranslate material-icons"
                role="presentation"
              >
                search
              </i>
            </div>
          </div>
          <!---->
        </div>
        <!---->
      </label>
      <button
        class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-primary q-btn--actionable q-focusable q-hoverable q-ml-md"
        data-test="refresh-button"
        tabindex="0"
        type="button"
      >
        <span
          class="q-focus-helper"
        />
        <span
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="q-icon notranslate material-icons"
            role="img"
          >
            restart_alt
          </i>
        </span>
        <transition-stub>
          <!---->
        </transition-stub>
      </button>
      <hr
        aria-orientation="vertical"
        class="q-separator q-separator--vertical q-separator--vertical-inset"
        style="margin-right: 8px; margin-left: 8px;"
      />
      <button
        class="q-btn q-btn-item non-selectable no-outline q-btn--unelevated q-btn--rectangle bg-green text-white q-btn--actionable q-focusable q-hoverable"
        data-test="add-button"
        tabindex="0"
        type="button"
      >
        <span
          class="q-focus-helper"
        />
        <span
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="q-icon on-left notranslate material-icons"
            role="presentation"
          >
            post_add
          </i>
          <div>
            ADD POST
          </div>
        </span>
      </button>
      
    </div>
    <div
      class="q-table__middle scroll"
    >
      <table
        class="q-table"
      >
        <thead>
          <tr
            class="q-tr"
          >
            
            <th
              class="text-left sortable text-primary"
            >
              Title
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left"
                role="presentation"
              >
                arrow_upward
              </i>
            </th>
            <th
              class="text-left text-primary"
            >
              Author
            </th>
            <th
              class="text-right sortable text-primary"
            >
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--right"
                role="presentation"
              >
                arrow_upward
              </i>
              Status
            </th>
            <th
              class="text-right sortable text-primary"
            >
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--right"
                role="presentation"
              >
                arrow_upward
              </i>
              Favorite
            </th>
            <th
              class="text-right sortable text-primary"
            >
              <i
                aria-hidden="true"
                class="q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--right"
                role="presentation"
              >
                arrow_upward
              </i>
              Views
            </th>
            <th
              class="text-center text-primary"
            >
              Actions
            </th>
            
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
          >
            <td
              class="text-left"
              style="width: 30%;"
            >
              first_title
            </td>
            <td
              class="text-left"
            >
              name
            </td>
            
            <td
              class="q-td q-table--col-auto-width text-right"
            >
              draft
            </td>
            
            
            <td
              class="q-td q-table--col-auto-width text-right"
            >
              usual
            </td>
            
            
            <td
              class="q-td q-table--col-auto-width text-right"
            >
              1
            </td>
            
            
            <td
              class="q-td q-table--col-auto-width text-center"
            >
              <div
                class="row flex-center q-gutter-x-sm no-wrap"
              >
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-accent q-btn--actionable q-focusable q-hoverable"
                  data-test="edit-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      edit
                    </i>
                    <!---->
                  </span>
                </button>
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-red q-btn--actionable q-focusable q-hoverable"
                  data-test="delete-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      delete
                    </i>
                    <!---->
                  </span>
                </button>
              </div>
            </td>
            
          </tr>
          <tr
            class=""
          >
            <td
              class="text-left"
              style="width: 30%;"
            >
              first_title
            </td>
            <td
              class="text-left"
            >
              name
            </td>
            
            <td
              class="q-td q-table--col-auto-width text-right"
            >
              draft
            </td>
            
            
            <td
              class="q-td q-table--col-auto-width text-right"
            >
              usual
            </td>
            
            
            <td
              class="q-td q-table--col-auto-width text-right"
            >
              1
            </td>
            
            
            <td
              class="q-td q-table--col-auto-width text-center"
            >
              <div
                class="row flex-center q-gutter-x-sm no-wrap"
              >
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-accent q-btn--actionable q-focusable q-hoverable"
                  data-test="edit-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      edit
                    </i>
                    <!---->
                  </span>
                </button>
                <button
                  class="q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle text-red q-btn--actionable q-focusable q-hoverable"
                  data-test="delete-button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="q-focus-helper"
                  />
                  <span
                    class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
                  >
                    <i
                      aria-hidden="true"
                      class="q-icon notranslate material-icons"
                      role="img"
                    >
                      delete
                    </i>
                    <!---->
                  </span>
                </button>
              </div>
            </td>
            
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="q-table__bottom row items-center justify-end"
    >
      <div
        class="q-table__separator col"
      />
      <div
        class="q-table__control"
      >
        <span
          class="q-table__bottom-item"
        >
          Records per page
        </span>
        <label
          class="q-field row no-wrap items-start q-field--borderless q-select q-field--auto-height q-select--without-input q-select--without-chips q-select--single q-field--float q-field--dense q-table__select inline q-table__bottom-item"
          for="f_7290fa0d-97a8-4a3b-98ba-58c0a03c5400"
        >
          <!---->
          <div
            class="q-field__inner relative-position col self-stretch"
          >
            <div
              class="q-field__control relative-position row no-wrap text-primary"
              tabindex="-1"
            >
              <div
                class="q-field__control-container col relative-position row no-wrap q-anchor--skip"
              >
                <div
                  class="q-field__native row items-center"
                >
                  <span>
                    5
                  </span>
                  <input
                    aria-autocomplete="none"
                    aria-controls="f_7290fa0d-97a8-4a3b-98ba-58c0a03c5400_lb"
                    aria-expanded="false"
                    aria-owns="f_7290fa0d-97a8-4a3b-98ba-58c0a03c5400_lb"
                    class="q-select__focus-target"
                    id="f_7290fa0d-97a8-4a3b-98ba-58c0a03c5400"
                    readonly=""
                    role="combobox"
                    tabindex="0"
                  />
                </div>
                <!---->
              </div>
              <div
                class="q-field__append q-field__marginal row no-wrap items-center q-anchor--skip"
              >
                <i
                  aria-hidden="true"
                  class="q-icon notranslate material-icons q-select__dropdown-icon"
                  role="presentation"
                >
                  arrow_drop_down
                </i>
              </div>
              <!---->
            </div>
            <!---->
          </div>
          <!---->
        </label>
      </div>
      <div
        class="q-table__control"
      >
        <span
          class="q-table__bottom-item"
        >
          1-5 of 20
        </span>
        <button
          aria-disabled="true"
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary disabled q-btn--dense"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              first_page
            </i>
          </span>
        </button>
        <button
          aria-disabled="true"
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary disabled q-btn--dense"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              chevron_left
            </i>
          </span>
        </button>
        <button
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary q-btn--actionable q-focusable q-hoverable q-btn--dense"
          tabindex="0"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              chevron_right
            </i>
          </span>
        </button>
        <button
          class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round text-primary q-btn--actionable q-focusable q-hoverable q-btn--dense"
          tabindex="0"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="q-icon notranslate material-icons"
              role="img"
            >
              last_page
            </i>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
